Unable to access bucket with restricted account in eu-central-1

Hi,

I have a somewhat peculiar problem. A user contributed a [tutorial](http://restic.readthedocs.io/en/latest/tutorials.html#setting-up-restic-with-amazon-s3) to setup a restricted account with S3 which can just access one bucket (and e.g. not create new buckets or access the console). Since I've updated minio-go in restic for https://github.com/minio/minio/issues/4275 I'm unable to create new repos via this restricted account (it worked before).

The error is (including the trace, built with minio-go 5d7ee332f62e83d36beba669be671801180abb89):
<details>

```
$ restic -r s3:s3.amazonaws.com/restic-test-travis/x2 init

---------START-HTTP---------                                   
GET /restic-test-travis/?location= HTTP/1.1
Host: s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJSQEUXZSP56YBL6Q/20170512/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=19ac77e663ecaee24ab7832479152865fc71ac5a06bf33914dcca617b59c6fc7
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20170512T203950Z
Accept-Encoding: gzip

HTTP/1.1 403 Forbidden
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 12 May 2017 20:39:51 GMT
Server: AmazonS3
X-Amz-Id-2: 5jNddocga32YohOP6BpvdE2/OxKFm1gOSiznLvdhXz6/ux5R3dAq1qoNQLTGswFzNZkafUtSMNI=
X-Amz-Request-Id: 31AD0B0559C5E892

f3
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>31AD0B0559C5E892</RequestId><HostId>5jNddocga32YohOP6BpvdE2/OxKFm1gOSiznLvdhXz6/ux5R3dAq1qoNQLTGswFzNZkafUtSMNI=</HostId></Error>
0
---------END-HTTP---------
---------START-HTTP---------
HEAD / HTTP/1.1
Host: restic-test-travis.s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJSQEUXZSP56YBL6Q/20170512/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=d87e750f8de5e6d21e8ceffca75b095eed69ff43744a1ac3f3e3a12029b70855
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170512T203951Z

HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 12 May 2017 20:39:51 GMT
Server: AmazonS3
X-Amz-Bucket-Region: eu-central-1
X-Amz-Id-2: 0qQw45YtBWvgSUm5HFoQ8UJ2w/XWn/KRpZyg6xj5DDMC7kPHhOFxVnIfM6OozkqvHOrVojoU49k=
X-Amz-Request-Id: AE1F2FC4AA4843B1

---------END-HTTP---------
create backend at s3:s3.amazonaws.com/restic-test-travis/x2 failed: client.BucketExists: 400 Bad Request
```

</details>

I've done a `git bisect` from b1674741d196d5d79486d7c1645ed6ded902b712 (good) to 5d7ee332f62e83d36beba669be671801180abb89 (bad), and the commit that caused it was:
```diff
fd942284fe190615d098af0b63e1698ef0969df1 is the first bad commit                                          
commit fd942284fe190615d098af0b63e1698ef0969df1                                                           
Author: Harshavardhana <harsha@minio.io>                                                                  
Date:   Tue Apr 4 11:35:20 2017 -0700                                                                     
                                                                                                          
    api: Check for Code 'InvalidRegion' for retrying with server Region. (#639)                           
                                                                                                          
diff --git a/api.go b/api.go
index 2beba77..e971721 100644
--- a/api.go
+++ b/api.go
@@ -543,9 +543,9 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 
        // For errors verify if its retryable otherwise fail quickly.
        errResponse := ToErrorResponse(httpRespToErrorResponse(res, metadata.bucketName, metadata.objectNa
-       // Bucket region if set in error response, we can retry the
-       // request with the new region.
-       if errResponse.Region != "" {
+       // Bucket region if set in error response and the error code dictates invalid region,
+       // we can retry the request with the new region.
+       if errResponse.Code == "InvalidRegion" && errResponse.Region != "" {
            c.bucketLocCache.Set(metadata.bucketName, errResponse.Region)
            continue // Retry.
        }
```

On the other hand, a successful trace (with the commit right before that one) is:
<details>

```
---------START-HTTP---------                               
GET /restic-test-travis/?location= HTTP/1.1
Host: s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.0.4
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170512/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20170512T203041Z
Accept-Encoding: gzip

HTTP/1.1 403 Forbidden
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 12 May 2017 20:30:41 GMT
Server: AmazonS3
X-Amz-Id-2: wFRgmf4iQpCw/spDBjf3kh6pZPCQ91bnrvBZGMs29ZqD3fLJfaHnoh0qtB1LphAv09Z67UKy0F8=
X-Amz-Request-Id: 9D1AEFF328416280

f3
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>9D1AEFF328416280</RequestId><HostId>wFRgmf4iQpCw/spDBjf3kh6pZPCQ91bnrvBZGMs29ZqD3fLJfaHnoh0qtB1LphAv09Z67UKy0F8=</HostId></Error>
0
---------END-HTTP---------
---------START-HTTP---------
HEAD / HTTP/1.1
Host: restic-test-travis.s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.0.4
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170512/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20170512T203041Z

HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 12 May 2017 20:30:41 GMT
Server: AmazonS3
X-Amz-Bucket-Region: eu-central-1
X-Amz-Id-2: yixxZF7DxK/3bJL9qlkrgHzjG7qDha2VrAtnhNLjezsoywZcjoxs9IsGhMN/wepfxOg++ekxou4=
X-Amz-Request-Id: 1C93A9104ED73A7D

---------END-HTTP---------
---------START-HTTP---------
HEAD / HTTP/1.1
Host: restic-test-travis.s3-eu-central-1.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.0.4
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170512/eu-central-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20170512T203041Z

HTTP/1.1 200 OK
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 12 May 2017 20:30:43 GMT
Server: AmazonS3
X-Amz-Bucket-Region: eu-central-1
X-Amz-Id-2: NT+/zXZ+hTi+iAKapHhlHByQ4a1I+ifFo3vbAdQBZ4Q3NEoyxgDxBIxqbHgmC/oWHPESv8S8mDc=
X-Amz-Request-Id: AB5769CE0F7D8CA6

---------END-HTTP---------
```

</details>

It seems to me that testing for `errResponse.Code == "InvalidRegion"` is not enough: It should also test for a response code of `400 Bad Request` when a region is present, and try that. I've verified that the following patch works:
```diff
diff --git a/api.go b/api.go
index 9cd2b6d..6beb445 100644
--- a/api.go
+++ b/api.go
@@ -556,7 +556,7 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 		// Bucket region if set in error response and the error
 		// code dictates invalid region, we can retry the request
 		// with the new region.
-		if errResponse.Code == "InvalidRegion" && errResponse.Region != "" {
+		if errResponse.Region != "" && (errResponse.Code == "InvalidRegion" || errResponse.Code == "400 Bad Request") {
 			c.bucketLocCache.Set(metadata.bucketName, errResponse.Region)
 			continue // Retry.
 		}
```

Maybe this also needs to be added to `api-put-bucket.go`?

FYI: I'm not sure if `errResponse.Code` will ever by equal to `InvalidRegion`, since in all cases I observed the string always started with the numeric HTTP response code and it also contained spaces...

Hm, this isn't solved completely, the program in 
https://gist.github.com/180efcf728398d991a1fa9c903a12f54 demonstrates it. It should stat and then upload ~120k of random data to a file in a bucket in region eu-central-1, but only creates an empty file (unrestricted user):
```
$ BUCKET=restic-dev-an go run main.go
---------START-HTTP---------
GET /restic-dev-an/?location= HTTP/1.1
Host: s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJCAOQML73FTJQFCA/20170513/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=a48fffc8eae41fa8032b4fa64abd23f18078a18eb67ef68055f499f03804c359
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170513T123113Z
Accept-Encoding: gzip

HTTP/1.1 200 OK
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Sat, 13 May 2017 12:31:14 GMT
Server: AmazonS3
X-Amz-Id-2: YXW0syryPA5XgxkrieTc1PHHQ6XsQkrYXA/InflCLNd5wWsxEQQmU3arC+fNXyvmAFQM4OIA7oE=
X-Amz-Request-Id: 73B06DAB9B274E36
---------END-HTTP---------
---------START-HTTP---------
HEAD /test-file-2596996162 HTTP/1.1
Host: restic-dev-an.s3-eu-central-1.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJCAOQML73FTJQFCA/20170513/eu-central-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=c02c5ccab8b20e78e5ef7ebaba42e614ac138d8a653536ea63c877ac3279f992
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170513T123113Z

Accept-Ranges: bytes
Content-Length: 0
Content-Type: binary/octet-stream
Date: Sat, 13 May 2017 12:31:14 GMT
Etag: "d41d8cd98f00b204e9800998ecf8427e"
Last-Modified: Sat, 13 May 2017 12:29:32 GMT
Server: AmazonS3
X-Amz-Id-2: 3nLE2t3YdUzZGPpJeyyDWjxOIpbz+/dLWcyJ1RplISm6Md5NSPCRXxc5w4zIrrUX1JQv3hdP4bQ=
X-Amz-Request-Id: 2334C8947A9F78CF
---------END-HTTP---------
stat: {d41d8cd98f00b204e9800998ecf8427e test-file-2596996162 2017-05-13 12:29:32 +0000 UTC 0 binary/octet-stream map[] { }  <nil>}
---------START-HTTP---------
PUT /test-file-2596996162 HTTP/1.1
Host: restic-dev-an.s3-eu-central-1.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Content-Length: 0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJCAOQML73FTJQFCA/20170513/eu-central-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=0bd1e7ec06716b575333d3c0ca0f2b191124fbb1e42b655ccb79a34a93ef62d8
Content-Md5: 1B2M2Y8AsgTpgAmY7PhCfg==
Content-Type: binary/octet-stream
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170513T123113Z
Accept-Encoding: gzip

Content-Length: 0
Date: Sat, 13 May 2017 12:31:14 GMT
Etag: "d41d8cd98f00b204e9800998ecf8427e"
Server: AmazonS3
X-Amz-Id-2: V8t0kl2BqjLF54OdfrtV48VVhWga/yTobpyUlAjoULJ6YHiveNogyMDmIchAq5q0vxIy/6jA1Ko=
X-Amz-Request-Id: A1E4821D3E411422
---------END-HTTP---------


wrote 0 bytes to test-file-2596996162 err <nil>
```

With the restricted user, it returns an error:
```
$ BUCKET=restic-test-travis go run main.go
---------START-HTTP---------                                                                                                                     
GET /restic-test-travis/?location= HTTP/1.1
Host: s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJSQEUXZSP56YBL6Q/20170513/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=dcbcee905db587e7501ed36e110dff595c4abbad73b20feec77e454c14ec1f18
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170513T123118Z
Accept-Encoding: gzip

HTTP/1.1 403 Forbidden
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Sat, 13 May 2017 12:31:18 GMT
Server: AmazonS3
X-Amz-Id-2: M6svkcVP9cFy/3xUbYcsNsWDp8ROtggNRltoMXiGTHf4SEjQQOOXPgEsffxHnIbkqNID4dRRzqs=
X-Amz-Request-Id: CA22CE3FD83BA6CC

f3
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>CA22CE3FD83BA6CC</RequestId><HostId>M6svkcVP9cFy/3xUbYcsNsWDp8ROtggNRltoMXiGTHf4SEjQQOOXPgEsffxHnIbkqNID4dRRzqs=</HostId></Error>
0
---------END-HTTP---------
---------START-HTTP---------
HEAD /test-file-2596996162 HTTP/1.1
Host: restic-test-travis.s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/2.1.0
Authorization: AWS4-HMAC-SHA256 Credential=AKIAJSQEUXZSP56YBL6Q/20170513/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=7f5d5a2387d00871b5de590b38d396652b8d9d8be1f1ac78be48af88604a7376
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170513T123118Z

HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Sat, 13 May 2017 12:31:17 GMT
Server: AmazonS3
X-Amz-Id-2: VDmZ3S0f4POwaGhUS9vAZzaxmB0+CUkVOibLV6pEwm76kkgm+V3fErLj5WAnIm0CCTsdCPUXxpI=
X-Amz-Request-Id: 89922C1E15AAFF6B

---------END-HTTP---------
panic: 400 Bad Request

goroutine 1 [running]:
main.main()
	/home/fd0/tmp/main.go:49 +0x7b1
exit status 2
```

When I remove the stat() call, there's no error in both cases but the file is empty (and `PutObject()` returns zero bytes).

Did I configure something wrong with the restricted user? I discovered that the fix does not resolve this issue completely because I'm reworking the restic integration tests to run against a Minio instance and against AWS S3 directly. One of my tests work for the unrestricted user that I'm using for development, but fails with an unexpected error message for the restricted user that I'm planning to expose to Travis:
```
    --- FAIL: TestBackendS3/Save (0.82s)
    	tests.go:361: client.PutObject: EOF
```

The code is in the restic repo at https://github.com/restic/restic in the master branch (I accidentally also pushed your fix for minio @harshavardhana). You can run the integration tests against S3 like this (replace BUCKET with a suitable bucket name):
```
$ export RESTIC_TEST_S3_KEY=$AWS_ACCESS_KEY_ID;
$ export RESTIC_TEST_S3_SECRET=$AWS_SECRET_ACCESS_KEY;
$ export RESTIC_TEST_S3_REPOSITORY=s3:s3.amazonaws.com/BUCKET;
$ GOPATH="$PWD:$PWD/vendor" go test -v -run 'TestBackendS3/Save$' restic/backend/s3
```

@fd0 able to reproduce the problem 

```
$ GOPATH="$PWD:$PWD/vendor" go test -v -run 'TestBackendS3/Save$' restic/backend/s3
=== RUN   TestBackendS3
=== RUN   TestBackendS3/Save
--- FAIL: TestBackendS3 (45.68s)
	s3_test.go:258: run tests
    --- FAIL: TestBackendS3/Save (37.78s)
    	tests.go:359: client.PutObject: EOF
FAIL
exit status 1
FAIL	restic/backend/s3	45.691s
```

An unrestricted user tests run fine.

```
 GOPATH="$PWD:$PWD/vendor" go test -v -run 'TestBackendS3/Save$' restic/backend/s3
=== RUN   TestBackendS3
=== RUN   TestBackendS3/Save
--- PASS: TestBackendS3 (75.93s)
	s3_test.go:258: run tests
    --- PASS: TestBackendS3/Save (69.73s)
PASS
ok  	restic/backend/s3	75.943s
```

With restricted user fails. Because it should fail. 

```
--- FAIL: TestBackendS3 (51.10s)
	s3_test.go:258: run tests
    --- FAIL: TestBackendS3/Save (44.58s)
    	tests.go:359: client.PutObject: EOF
FAIL
exit status 1
FAIL	restic/backend/s3	51.111s
```

After reflecting on the incoming Reader() it can be known that 

```
test.errorCloser
```

is of following type which doesn't implement Len(), Size(), or any other compatible implementations required by function `getReaderSize()`  https://github.com/minio/minio-go/blob/master/api-put-object.go#L55 

Since these methods don't exist minio-go sets the reader size to be '-1' i.e starts uploading as multipart, since all multipart APIs are disabled with restricted user it fails. In S3 no PUT operation is allowed without knowing the size of the object i.e Content-Length. It is a requirement and fallback style which minio-go follows to be able to somehow upload the file if possible. But when all else fails it fails. 

It should be noted that some tests succeed which are `*bytes.Reader` which implement Len() and Size().  

So here is the fix 

```diff
diff --git a/src/restic/backend/test/tests.go b/src/restic/backend/test/tests.go
index 4519ee2..a643f65 100644
--- a/src/restic/backend/test/tests.go
+++ b/src/restic/backend/test/tests.go
@@ -277,7 +277,12 @@ func BackendTestLoad(t testing.TB, s *Suite) {
 
 type errorCloser struct {
        io.Reader
-       t testing.TB
+       size int64
+       t    testing.TB
+}
+
+func (ec errorCloser) Size() int64 {
+       return ec.size
 }
 
 func (ec errorCloser) Close() error {
@@ -354,7 +359,7 @@ func BackendTestSave(t testing.TB, s *Suite) {
 
        // wrap the tempfile in an errorCloser, so we can detect if the backend
        // closes the reader
-       err = b.Save(h, errorCloser{t: t, Reader: tmpfile})
+       err = b.Save(h, errorCloser{t: t, size: int64(len(data)), Reader: tmpfile})
        if err != nil {
                t.Fatal(err)
        }
```

--- a/api-put-bucket.go
+++ b/api-put-bucket.go
@@ -91,7 +91,7 @@ func (c Client) MakeBucket(bucketName string, location string) (err error) {
 		if resp.StatusCode != http.StatusOK {
 			err := httpRespToErrorResponse(resp, bucketName, "")
 			errResp := ToErrorResponse(err)
-			if errResp.Code == "InvalidRegion" && errResp.Region != "" {
+			if resp.StatusCode == http.StatusBadRequest && errResp.Region != "" {
 				// Fetch bucket region found in headers
 				// of S3 error response, attempt bucket
 				// create again.

--- a/api.go
+++ b/api.go
@@ -556,7 +556,7 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 		// Bucket region if set in error response and the error
 		// code dictates invalid region, we can retry the request
 		// with the new region.
-		if errResponse.Code == "InvalidRegion" && errResponse.Region != "" {
+		if res.StatusCode == http.StatusBadRequest && errResponse.Region != "" {
 			c.bucketLocCache.Set(metadata.bucketName, errResponse.Region)
 			continue // Retry.
 		}

